### ISSUE TITLE
How to list SFTP files recursively, efficiently

### ISSUE THREAD

**Issue by josephernest** [ronf/asyncssh#332]
I have a SFTP server on my phone with ~ 20k files. This works but it's very slow (280 seconds):

```
async def main():
    async with asyncssh.connect(HOST, port=PORT, username=USER, password='') as conn:
        async with conn.start_sftp_client() as sftp:
            for f in await sftp.glob('music/**/*'):
                print(f)
asyncio.run(main())
```

This is better (~ 60 seconds):

```
tasks = set()
FILES = []

async def list_recursive(sftp, remotedir):
    print(remotedir)
    for f in await sftp.readdir(remotedir):
        if f.longname[0] == 'd':  # directory
            tasks.add(asyncio.create_task(list_recursive(sftp, remotedir + '/' + f.filename)))
        else:
            FILES.append(f.filename)

async def main():
    async with asyncssh.connect(HOST, port=PORT, username=USER, password='') as conn:
        async with conn.start_sftp_client() as sftp:
            await list_recursive(sftp, 'music')
            print('hello')
            await asyncio.wait(tasks)

asyncio.run(main())

print(len(FILES))
```

I'm not sure about this: are we syre all the tasks are started properly (even if there are subdir, subdirs, subsubdirs, etc.) before             `await asyncio.wait(tasks)` happens?

**Do you see a way to improve the file tree walking with asynchronous methods @ronf?**

**Comment by ronf** [comment]
I think you're on the right track in terms of allowing multiple directories to be listed in parallel to improve the speed, assuming there are enough separate subdirectories to make it worth the effort. However, I would definitely be concerned about trying to use a single wait() call on a list that might still be getting updated due to new tasks being created.

While it can't be used as-is for directory listings, you might want to take a look at the SFTPParallelIO class in sftp.py in AsyncSSH for an example of how to schedule multiple requests in parallel and wait for all of them to complete. It uses a "_pending" set which is similar to your "tasks" variable here, but it calls asyncio.wait() with the "FIRST_COMPLETED" argument, and has that return both a list of tasks which are done and tasks which are still pending. After processing the tasks which are done, it then schedules additional tasks if needed and loops back around until the "_pending" list is completely empty. For what you need here, you don't need some of the abstractions of letting child classes implement some of the methods, but the basic loop structure in the run() function might be useful as a model.

**Comment by josephernest** [comment]
Thanks @ronf for your answer.

> I would definitely be concerned about trying to use a single wait() call on a list that might still be getting updated due to new tasks being created.

Totally. It's exactly the conceptual problem I have with `asyncio`. I'm sure there is an elegant solution to this, but I don't know it yet. Here I posted this linked question: https://stackoverflow.com/questions/65136464/how-to-use-asyncio-wait-on-an-growing-set-of-tasks

> . It uses a "_pending" set which is similar to your "tasks" variable here, but it calls asyncio.wait() with the "FIRST_COMPLETED" argument, and has that return both a list of tasks which are done and tasks which are still pending. After processing the tasks which are done, it then schedules additional tasks if needed and loops back around until the "_pending" list is completely empty.

Just to be sure, how would it look here (even in pseudo code it's ok) with `list_recursive`?

**Comment by josephernest** [comment]
Based on [this answer](https://stackoverflow.com/questions/65136464/how-to-use-asyncio-wait-on-an-growing-set-of-tasks/65139941#65139941), here seems to be a nice solution:

```
FILES = []

async def list_recursive(sftp, remotedir):
    tasks = set()
    print(remotedir)
    for f in await sftp.readdir(remotedir):
        if f.longname[0] == 'd':  # directory
            tasks.add(asyncio.create_task(list_recursive(sftp, remotedir + '/' + f.filename)))
        else:
            FILES.append(f.filename)
    if tasks:
        await asyncio.wait(tasks)

async def main():
    async with asyncssh.connect(HOST, port=PORT, username=USER, password='') as conn:
        async with conn.start_sftp_client() as sftp:
            print('connected')
            await list_recursive(sftp, 'music')
    
asyncio.run(main())
```

Each `list_recursive` call has its *own* set of child tasks, and it `await` for them at the end of the `list_recursive` call, i.e. each call is responsible for its own sub-tasks, and there is no need for a global set.

**Comment by ronf** [comment]
Good find.

Having each task wait() on on all the subtasks it creates seems like a good approach if you are always starting from a single directory. If you had a list of directories to search, you'd need a top-level list for that which you called wait() on, but even then you could still let any subdirectory tasks be picked up by the parent task.

**Comment by ronf** [comment]
> Just to be sure, how would it look here (even in pseudo code it's ok) with list_recursive?

What you ended up with looks fine, but for completeness here's the existing loop in SFTPParallelIO that I mentioned:

```
        try:
            await self.start()

            self._start_tasks()

            while self._pending:
                done, self._pending = await asyncio.wait(
                    self._pending, return_when=asyncio.FIRST_COMPLETED)

                exceptions = []

                for task in done:
                    exc = task.exception()

                    if exc and not isinstance(exc, SFTPEOFError):
                        exceptions.append(exc)

                if exceptions:
                    for task in self._pending:
                        task.cancel()

                    raise exceptions[0]

                self._start_tasks()

            return await self.finish()
        finally:
            await self.cleanup()
```

The start_tasks() call could be replaced with something that did the create_task() on each directory you find, with the version at the top changed to start off the process on the initial directory (or directories) you want to scan. The start() and finish() calls are just placeholders to allow child classes of this to do some initialization and cleanup. 

You wouldn't need the SFTPEOFError check, but you might want to think about adding code to try and retrieve any exceptions which occur. You can then decide if you want to cancel all the other outstanding tasks or let them all complete and just ignore the directories which you get an error on.

**Comment by ronf** [comment]
You might want to take a look at the new scandir() I added as part of the work in #334 to help speed up directory listings. You could potentially start parallel fetches of subdirectories without first fetching the entire contents of a parent directory. All you need to do is switch from a "for" loop on the return value of readdir() to an "async for" on the return value of the new scandir() function.

**Comment by josephernest** [comment]
Thank you @ronf for this addition that looks nice!

1) Is it possible to `pip install asyncssh==testing` or something like this to be able to quickly test it here? What's the appropriate way to do it with `pip`? Such that I can `pip install asyncssh` the stable version again later.

2) Up to now I used this:

```
async def list_recursive(sftp, remotedir):
    tasks = set()
    for f in await sftp.readdir(remotedir):
        print(f.filename)  # here in reality, I did the processing to do with these files
        remotepath = remotedir + "/" + f.filename
        if stat.S_ISDIR(f.attrs.permissions):
            tasks.add(asyncio.create_task(list_recursive(sftp, remotepath)))
    if tasks:
        await asyncio.wait(tasks)
```
By what would you replace this, would you have an example in mind? (I'm not familiar with `async for`)

**Comment by ronf** [comment]
I think all you'd need to do here is to replace "for f in await sftp.readdir(remotedir)" with "async for f in sftp.scandir(remotedir)".

**Comment by ronf** [comment]
Regarding installing the "develop" version, you should be able to do something like:

    pip install git+https://github.com/ronf/asyncssh@develop

However, since the "develop" version generally has the same version number as the latest released version, you may need to "pip uninstall asyncssh" before doing that or add a "--force-reinstall" to the above if you already have the latest version installed.

**Comment by josephernest** [comment]
Thank you for your answers!

Strangely, if I compare these two methods:

```
import asyncio, asyncssh, time, stat

async def list_recursive1(sftp, remotedir):
    tasks = set()
    for f in await sftp.readdir(remotedir):
        print(f.filename)
        remotepath = remotedir + "/" + f.filename
        if stat.S_ISDIR(f.attrs.permissions):
            tasks.add(asyncio.create_task(list_recursive1(sftp, remotepath)))
    if tasks:
        await asyncio.wait(tasks)

async def list_recursive2(sftp, remotedir):
    async for f in sftp.scandir(remotedir):
        print(f.filename)
        remotepath = remotedir + "/" + f.filename
        if stat.S_ISDIR(f.attrs.permissions):
            await list_recursive2(sftp, remotepath)     # can we do better than this await here?

async def main():
    global t0
    async with asyncssh.connect('192.168.1.18', port=2222, username='ssh', password='', known_hosts=None) as conn:
        async with conn.start_sftp_client() as sftp:
            t0 = time.time()                        # listing ~ 20k files / 1k folder on my Android phone (running a SSH server)
            await list_recursive1(sftp, 'music')    # 59.9 seconds
            # await list_recursive2(sftp, 'music')    # 74.8 seconds
asyncio.run(main())         
```

then `list_recursive1` with a `set()` of tasks, `readdir`,  seems faster than `list_recursive2` with `async for ... scandir`  (59.9 seconds vs 74.8 seconds). Do you have an idea of what could be improved in `list_recursive2`, that lists all files recursively in subfolders, etc.?

**Comment by ronf** [comment]
The gain here will most likely only show up if you have directories with a large number of files (hundreds or thousands). If you've got 20k files spread across 1k directories, the entire contents of each directory is probably returned in a single SFTP request, so you're probably just paying extra context switching overhead for the "async for" without actually getting any more parallelism.

I saw something similar with rmtree() in some of the testing I did. With 10k files in 10 directories, the scandir() version was about half a second faster (12.26 sec vs 12.77 sec). However, with 10k files in 1k directories, it was about 0.1 second slower (18.26 sec vs 18.16 sec).

Doing a shutil.rmtree() on my local machine on the same set of files completes in like 0.55 seconds, so clearly there's a lot of overhead introduced by SFTP. I'm not sure how much of that is in listing the files vs. actually doing the 10,000 unlink() calls in the rmtree case, but I'll play with it a bit more to see if there's anything else that can be done there.

Did you try leaving in the tasks.add() and the wait() in conjunction with the "async for"? For large directories, that's what I would expect to get the most gain, as you'd begin doing the listing of child directories even before completing a listing on the parent directory. However, if the parent directories all have a small number of files, that'll probably just make it slower.

**Comment by ronf** [comment]
One thing I just noticed about your list_recursive functions -- you don't seem to be filtering out '.' and '..' directories. This may be causing it to visit more than just the files you intended, though I would have thought in that case that it would have never completed. Perhaps your SFTP server isn't including these entries in the response.

When I tried it with OpenSSH as a server, I had to add the following lines right after the "for f" (or "async for f") statement:

```python
        if f.filename in ('.', '..'):
            continue
```

When I did this and tested it on 20 directories with 1000 files each, it ran in 0.42 seconds here (going to localhost). Testing list_recusive2, it was definitely slower (0.66 seconds), but the following was slightly faster (0.40 seconds):

```python
async def list_recursive3(sftp, remotedir):
    tasks = set()
    async for f in sftp.scandir(remotedir):
        if f.filename in ('.', '..'):
            continue

        print(f.filename)
        remotepath = remotedir + "/" + f.filename
        if stat.S_ISDIR(f.attrs.permissions):
            tasks.add(asyncio.create_task(list_recursive3(sftp, remotepath)))
    if tasks:
        await asyncio.wait(tasks)
```

This is what I suggested above.

I expect an SFTP server on a phone might be a bit slower for all of these tests, but you may want to confirm just how many files it is returning and whether it is possibly returning files outside of the tree you intended.

I also tried this with 2000 directories of 10 files each, but I ran into SFTP failures with versions 1 & 3 here. I think the SFTP server I was using had a limit on the number of file handles it lets a client open and the parallelism was too great. To make this work, there might need to be a limit on the number of tasks it starts at once, similar to what I mentioned above in the ParallelIO class where it adds new tasks only after other tasks complete. This is handled in the _start_tasks() function that is present in the subclasses.

I may actually need to do something like this in the new rmtree() code I just added, to prevent it from trying to list too many directories at once.

**Comment by ronf** [comment]
Ok - here's how you can bound the number of parallel tasks:

```python
import asyncio, asyncssh, stat, sys, time

_MAX_SFTP_REQUESTS = 128

async def list_recursive3(sftp, sem, remotedir):
    tasks = set()

    async with sem:
        async for f in sftp.scandir(remotedir):
            if f.filename in ('.', '..'):
                continue

            print(f.filename)
            remotepath = remotedir + "/" + f.filename
            if stat.S_ISDIR(f.attrs.permissions):
                tasks.add(asyncio.create_task(list_recursive3(sftp, sem, remotepath)))
        if tasks:
            await asyncio.wait(tasks)

async def main():
    async with asyncssh.connect('localhost') as conn:
        async with conn.start_sftp_client() as sftp:
            sem = asyncio.Semaphore(_MAX_SFTP_REQUESTS)
            t0 = time.time()
            await list_recursive3(sftp, sem, 'tree')
            print(time.time() - t0, file=sys.stderr)

asyncio.run(main())         
```

Approach 2 took about 2.66 seconds for me on 2000 directories with 10 files each. Switching to approach 3 dropped that to 1.86 seconds. This was comparable to approach 1 when it was modified to also add the semaphore. That took 1.87 seconds for me, which is not a surprise -- directories with only a few files will return all of them in a single SFTP response, so you don't get extra parallelism from scandir() over reader() in that case.

**Comment by ronf** [comment]
The new scandir() method is now available in AsyncSSH 2.5.0, as is the rmtree() change to limit the number of parallel tasks it will start at once.
